CoreWebServer/src/au/gov/asd/tac/constellation/webserver/help/SphinxHelpDisplayer.java

```diff
@@ -0,0 +1,322 @@
+import java.util.zip.ZipInputStream;
+import org.openide.awt.NotificationDisplayer;
+import org.openide.util.Exceptions;
+import org.openide.util.HelpCtx;
+import org.openide.util.lookup.ServiceProvider;
+
+/**
+ *
+ * @author algol
+ */
+@ServiceProvider(service = HelpCtx.Displayer.class, position = 9999)
+public class SphinxHelpDisplayer implements HelpCtx.Displayer {
+
+    private static final Logger LOGGER = Logger.getLogger(SphinxHelpDisplayer.class.getName());
+
+    // This system property controls where the help is displayed from.
+    //
+    // * If it is undefined, help is read from getResourceAsStream().
+    // * If it is defined as a (zip) file, help is read from that zip file.
+    // * If it is defined as a web URL, help is read from that URL.
+    //
+    // Define it in this module's project.properties file or another
+    // appropriate place as (for example):
+    //
+    // run.args.extra="-J-Dconstellation.help=D:/tmp/help.zip"
+    // run.args.extra="-J-Dconstellation.help=http://localhost:8000"
+    // run.args.extra="-J-Dconstellation.help=https://example.com"
+    //
+    //
+    private static final String CONSTELLATION_HELP = "constellation.help";
+
+    // The help zip must contain a text file mapping helpIds to page paths.
+    //
+    private static final String HELP_MAP = "help_map.txt";
+
+    // The name of the internal help.zip resource.
+    //
+    private static final String HELP_ZIP = "help.zip";
+
+    // Maximum size of buffer for reading help files.
+    //
+    private static final int BUFSIZ = 10 * 1024 * 1024;
+
+    private static String helpSource;
+
+    // The mapping of helpIds to page paths.
+    //
+    private static Map<String, String> helpMap;
+
+    /**
+     * Read an entry from the zip file HELP_ZIP stored as an internal resource.
+     * <p>
+     * Because we only get access via getResourceAsStream(), we have to iterate
+     * through the entries to get the right one.
+     *
+     * @param filepath The zip entry to get.
+     * @param out The OutputStream to write the entry's contents to.
+     *
+     * @throws IOException
+     */
+    private static void copyFromZipResource(final String filepath, final OutputStream out) throws IOException {
+        final InputStream in = SphinxHelpDisplayer.class.getResourceAsStream(HELP_ZIP);
+        if (in != null) {
+            final ZipInputStream zin = new ZipInputStream(in);
+            ZipEntry entry;
+            while ((entry = zin.getNextEntry()) != null) {
+                final String name = entry.getName();
+                if (name.equals(filepath)) {
+                    final byte[] buf = zin.readNBytes(BUFSIZ);
+                    zin.closeEntry();
+                    zin.close();
+
+                    out.write(buf);
+
+                    return;
+                }
+            }
+
+            LOGGER.log(Level.WARNING, "Could not find entry '{0}' in resource {1}", new Object[]{HELP_MAP, HELP_ZIP});
```

Comment:
remove the single quotes from `'{0}'` as the quotes are actually discarding format specifier (according to sonar)

Reply:
thank you - fixed.